Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -0,0 +1,525 @@
+ */
+CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localizedString) {
+    if (localizedString == NULL) {
+        TraceError(TAG, L"The input parameter is invalid!");
+        return nil;
+    }
+
+    // Get the default locale for this user.
+    wstring localeName = __GetUserDefaultLocaleName();
+
+    uint32_t index = 0;
+    BOOL exists = false;
+
+    // If the default locale is returned, find that locale name, otherwise use "en-us".
+    RETURN_NULL_IF_FAILED(localizedString->FindLocaleName(localeName.c_str(), &index, &exists));
+    if (!exists) {
+        RETURN_NULL_IF_FAILED(localizedString->FindLocaleName(c_defaultUserLanguage, &index, &exists));
+    }
+
+    // If the specified locale doesn't exist, select the first on the list.
+    if (!exists) {
+        index = 0;
+    }
+
+    // Get the string length.
+    uint32_t length = 0;
+    RETURN_NULL_IF_FAILED(localizedString->GetStringLength(index, &length));
+
+    // Get the string.
+    vector<wchar_t> wcharString = std::vector<wchar_t>(length + 1, 0);
+    RETURN_NULL_IF_FAILED(localizedString->GetString(index, wcharString.data(), length + 1));
+
+    // Strip out unnecessary null terminator
+    return (CFStringRef)CFAutorelease(
+        CFStringCreateWithCharacters(kCFAllocatorSystemDefault, reinterpret_cast<UniChar*>(wcharString.data()), length));
+}
+
+/**
+ * Helper method to retrieve font family names installed in the system.
+ *
+ * @return Immutable array of font family name strings that are installed in the system.
+ */
+CFArrayRef _DWriteGetFontFamilyNames() {
+    woc::unique_cf<CFMutableArrayRef> fontFamilyNames(CFArrayCreateMutable(kCFAllocatorSystemDefault, 0, &kCFTypeArrayCallBacks));
+
+    // Get the direct write factory instance
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_NULL_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    // Get the system font collection.
+    ComPtr<IDWriteFontCollection> fontCollection;
+    RETURN_NULL_IF_FAILED(dwriteFactory->GetSystemFontCollection(&fontCollection));
+
+    // Get the number of font families in the collection.
+    uint32_t count = 0;
+    count = fontCollection->GetFontFamilyCount();
+
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        RETURN_NULL_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        RETURN_NULL_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        CFStringRef name = _CFStringFromLocalizedString(familyNames.Get());
+        if (CFStringGetLength(name) == 0) {
+            TraceError(TAG, L"Failed to convert the localized string to wide string.");
+            continue;
+        }
+
+        CFArrayAppendValue(fontFamilyNames.get(), name);
+    }
+
+    return fontFamilyNames.release();
```

Comment:
While true, the "correct" fix here would be to return a +1 item with a different name because CFStyle things should not obey the same idea of +0 items. Those don't really exist in CF land. Unowned returns exist but that semantic is different that what is going on here. CFAutorelease *really* shouldn't be used if possible. since these are private functions I'd prefer a unique_cf return to correctly describe the *transfer* of ownership to the caller which is the pattern that is actually be attempted here with CFAutorelease being used to make the return work with a raw * which is just a silly Objective-C ism due to the bolt on nature of extra meaning to a * without an ability to in language convey that. Relying on naming conventions to express memory management is an error prone system that I'd like to not rely on ... 

---
In reply to: [84563998](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r84563998) [](ancestors = 84563998)